autogpts/autogpt/autogpt/agents/agent.py

```diff
@@ -64,157 +73,181 @@ class AgentConfiguration(BaseAgentConfiguration):
+            await self.foreach_components("get_messages")
         )
 
-    def on_before_think(self, *args, **kwargs) -> ChatPrompt:
-        prompt = super().on_before_think(*args, **kwargs)
+        prompt: ChatPrompt = await self.foreach_components(
+            "build_prompt",
+            messages,
+            self.commands,
+            self.state.task,
+            self.state.ai_profile,
+            directives,
+        )
 
         self.log_cycle_handler.log_count_within_cycle = 0
         self.log_cycle_handler.log_cycle(
-            self.ai_profile.ai_name,
+            self.state.ai_profile.ai_name,
             self.created_at,
             self.config.cycle_count,
             prompt.raw(),
             CURRENT_CONTEXT_FILE_NAME,
         )
-        return prompt
-
-    def parse_and_process_response(
-        self, llm_response: AssistantChatMessage, *args, **kwargs
-    ) -> Agent.ThoughtProcessOutput:
-        for plugin in self.config.plugins:
-            if not plugin.can_handle_post_planning():
-                continue
-            llm_response.content = plugin.post_planning(llm_response.content or "")
-
-        (
-            command_name,
-            arguments,
-            assistant_reply_dict,
-        ) = self.prompt_strategy.parse_response_content(llm_response)
-
-        # Check if command_name and arguments are already in the event_history
-        if self.event_history.matches_last_command(command_name, arguments):
-            raise DuplicateOperationError(
-                f"The command {command_name} with arguments {arguments} "
-                f"has been just executed."
+
+        logger.debug(f"Executing prompt:\n{dump_prompt(prompt)}")
+        output = await self.complete_and_parse(prompt)
+        self.config.cycle_count += 1
+
+        return output
+
+    @retry()
+    async def complete_and_parse(
+        self, prompt: ChatPrompt, exception: Optional[Exception] = None
+    ) -> ThoughtProcessOutput:
+        if exception:
+            prompt.messages.append(ChatMessage.system(f"Error: {exception}"))
+
+        response: AssistantChatMessage = (
+            await self.llm_provider.create_chat_completion_raw(
+                prompt.messages,
+                functions=(
+                    get_openai_command_specs(self.commands)
+                    if self.config.use_functions_api
+                    else []
+                ),
+                model_name=self.llm.name,
             )
+        )
+        result = ThoughtProcessOutput()
+        result: ThoughtProcessOutput = await self.foreach_components(
+            "parse_response", result, response
+        )
+
+        # Check if the command is valid, e.g. isn't duplicating a previous command
+        command = self.get_command(result.command_name)
+        if command:
+            is_valid, reason = command.is_valid(result.command_args)
+            if not is_valid:
+                raise InvalidOperationError(reason)
 
```

Comment:
What is this log cycle? Would modifying action history a bit to include more information allow to just remove this log cycle entirely? It needs to be move to a separate component otherwise.